src/HOI4World/Map/Railways.cpp

```diff
@@ -14,152 +16,221 @@ using HoI4::Railway;
-				{
-					newPaths.push(newPath);
-				}
-			}
-		}
+	std::priority_queue<std::vector<int>> new_paths;
+	new_paths.push({vic2_province_num});
+
+	while (!new_paths.empty())
+	{
+		std::vector<int> path = new_paths.top();
+		new_paths.pop();
 
-		std::vector<int> bestPotentialNewPath;
-		long potentialNewPathScore = 0;
-		for (const auto& potentialNewVic2ProvincePath: potentialNewVic2ProvincePaths)
+		for (const auto& new_path:
+			 GetNextPaths(path, valid_vic2_province_numbers, vic2_provinces, vic2_map_data, handled_provinces))
 		{
-			if (potentialNewVic2ProvincePath.size() == 2)
+			if (valid_vic2_province_numbers.contains(new_path.back()))
 			{
-				auto reversedPath = potentialNewVic2ProvincePath;
-				std::ranges::reverse(reversedPath);
-				if (!vic2ProvincePaths.contains(reversedPath))
-				{
-					vic2ProvincePaths.insert(potentialNewVic2ProvincePath);
-				}
-				continue;
+				potential_new_vic2_possible_paths.insert(new_path);
 			}
-
-			const int potentialLastProvinceNum = potentialNewVic2ProvincePath[potentialNewVic2ProvincePath.size() - 1];
-			const auto potentialLastProvinceItr = Vic2Provinces.find(potentialLastProvinceNum);
-			if (potentialLastProvinceItr == Vic2Provinces.end())
+			else
 			{
-				continue;
+				new_paths.push(new_path);
 			}
-			const int potentialLastProvincePopulation = potentialLastProvinceItr->second->getPopulation();
-			if (potentialLastProvincePopulation > potentialNewPathScore)
+		}
+	}
+
+	return potential_new_vic2_possible_paths;
+}
+
+
+int GetProvincePopulation(int province_num, const std::map<int, std::shared_ptr<Vic2::Province>>& vic2_provinces)
+{
+	const auto potential_last_province_itr = vic2_provinces.find(province_num);
+	if (potential_last_province_itr == vic2_provinces.end())
+	{
+		return 0;
+	}
+	return potential_last_province_itr->second->getPopulation();
+}
+
+
+std::vector<int> DetermineBestPotentialNewPath(const std::set<std::vector<int>>& potential_new_vic2_possible_paths,
+	 const std::map<int, std::shared_ptr<Vic2::Province>>& vic2_provinces,
+	 std::set<std::vector<int>>& vic2_possible_paths)
+{
+	std::vector<int> best_potential_new_path;
+	int32_t potential_new_path_score = 0;
+
+	for (const auto& potential_new_vic2_possible_path: potential_new_vic2_possible_paths)
+	{
+		if (potential_new_vic2_possible_path.size() == 2)
+		{
+			std::vector<int> reversed_path = potential_new_vic2_possible_path;
+			std::ranges::reverse(reversed_path);
+			if (!vic2_possible_paths.contains(reversed_path))
 			{
-				bestPotentialNewPath = potentialNewVic2ProvincePath;
-				potentialNewPathScore = potentialLastProvincePopulation;
+				vic2_possible_paths.insert(potential_new_vic2_possible_path);
 			}
+			continue;
+		}
+
+		const int potential_last_province_num = potential_new_vic2_possible_path.back();
+		const int potential_last_province_population = GetProvincePopulation(potential_last_province_num, vic2_provinces);
+		if (potential_last_province_population > potential_new_path_score)
```

Comment:
For example here, if one province was mountain and other was a flat, even unpopulated flat, go through that maybe and use trucks up the mountain? Just a thought.